docs migration: Container Lifecycle Hooks into Concepts

chenopis opened the pull request:
Refactor [`docs/user-guide/container-environment.md`](https://kubernetes.io/docs/user-guide/container-environment) to `docs/concepts/containers/container-lifecycle-hooks.md` and `docs/concepts/containers/container-environment-variables.md`

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes.github.io/2564)
<!-- Reviewable:end -->

chenopis commented:
@steveperry-53 Ok, this is ready for you to review at your convenience. Thx

steveperry-53 commented:
@chenopis, This topic covers more that what the title suggests. We should either make the title more general or limit the topic content to Container lifecycle hooks. I vote for limiting the topic content to Container lifecycle hooks. If you agree, then I think you should remove the "Container environment" section, and shorten the topic overview.

steveperry-53 commented:
@chenopis, Capitalize API objects like Container. I don't think "hook" is an API object, so it shouldn't be capitalized.

steveperry-53 commented on docs/concepts/workloads/containers/container-lifecycle-hooks.md:
```diff
@@ -0,0 +1,163 @@
+
+Container hooks provide information to the container about events in its management lifecycle.  
+For example, immediately after a container is started, it receives a `PostStart` hook.  
+Hooks are broadcast in the container with information about the lifecycle of the container.  
```
I think "with information about" is misleading. The hook handlers do not receive any arguments. When the Container's PostStart handler is called, it knows the Container was recently started; that's the only information it gets.

steveperry-53 commented on docs/concepts/workloads/containers/container-lifecycle-hooks.md:
```diff
@@ -0,0 +1,163 @@
+Container hooks provide information to the container about events in its management lifecycle.  
+For example, immediately after a container is started, it receives a `PostStart` hook.  
+Hooks are broadcast in the container with information about the lifecycle of the container.  
+They are different from the events provided by Docker or other systems which are output by the container.  
```
I think these sentences about Docker and output events take the reader on an unnecessary side trip. I would remove them.

steveperry-53 commented on docs/concepts/workloads/containers/container-lifecycle-hooks.md:
```diff
@@ -0,0 +1,163 @@
+
+This hook is called immediately before a container is terminated. 
+No parameters are passed to the handler. 
+This event handler is blocking and must complete before the call to delete the container is sent to the Docker daemon. 
```
I think this would be more clear if we phrased it something like: "must complete before Kubernetes deletes the Container."

steveperry-53 commented on docs/concepts/workloads/containers/container-lifecycle-hooks.md:
```diff
@@ -0,0 +1,163 @@
+This hook is called immediately before a container is terminated. 
+No parameters are passed to the handler. 
+This event handler is blocking and must complete before the call to delete the container is sent to the Docker daemon. 
+The `SIGTERM` notification sent by Docker is also sent. 
```
I think we can remove this sentence about Docker sending SIGTERM. It strikes me as an implementation detail.

steveperry-53 commented on docs/concepts/workloads/containers/container-lifecycle-hooks.md:
```diff
@@ -0,0 +1,163 @@
+No parameters are passed to the handler. 
+This event handler is blocking and must complete before the call to delete the container is sent to the Docker daemon. 
+The `SIGTERM` notification sent by Docker is also sent. 
+A more complete description of the termination behavior can be found in [Termination of Pods](/docs/user-guide/pods/#termination-of-pods).
```
Switch to active voice whenever you can. "You can find a more complete description ..."

steveperry-53 commented on docs/concepts/workloads/containers/container-lifecycle-hooks.md:
```diff
@@ -0,0 +1,163 @@
+
+### Hook handler execution
+
+When a management hook occurs, 
```
This could be simpler. "Immediately after a Container starts, Kubernetes calls the Container's PostStart handler."

steveperry-53 commented on docs/concepts/workloads/containers/container-lifecycle-hooks.md:
```diff
@@ -0,0 +1,163 @@
+When a management hook occurs, 
+the management system calls into any registered hook handlers in the container for that hook.  
+These hook handler calls are synchronous in the context of the pod containing the container. 
+This means that for a `PostStart` hook, 
```
This is hard to follow. "calls are synchronous ... This means ... asynchronously."

Consider something similar to the Discussion section in [Attaching Handlers to Container Lifecycle Events](https://kubernetes.io/docs/tasks/configure-pod-container/attach-handler-lifecycle-event/).

steveperry-53 commented on docs/concepts/workloads/containers/container-lifecycle-hooks.md:
```diff
@@ -0,0 +1,163 @@
+
+### Hook handler implementations
+
+Hook handlers are the method by which Hooks are exposed to Containers.  
```
This first sentence is a bit contorted. I would remove it.

steveperry-53 commented:
@chenopis, Does this PR need all 7 files? I don't understand why volumes.md and images.md are here.

steveperry-53 commented:
@chenopis, Back to you.

chenopis commented:
@steveperry-53 Ok, I'll break the topic into smaller parts and incorporate your feedback.Thanks!

steveperry-53 commented:
@chenopis, I don't think we need a separate topic for Container Environment. I think we could just remove that content and be fine. The part about FOO_SERVICE_HOST/PORT is obsolete (@pwittrock, correct?) and the rest is general information that I think is made clear throughout the docs.

steveperry-53 commented:
@chenopis, In What's next, add a link to [Attaching Handlers to Container Lifecycle Events](https://kubernetes.io/docs/tasks/configure-pod-container/attach-handler-lifecycle-event/).

chenopis commented:
@steveperry-53 Ok, thanks. Will do!

steveperry-53 commented:
cc: @devin-donnelly for the moving of user-guide/images.md. My guess it that this file won't just get moved as a single topic under Concepts. It looks like some of the content should go under Setup Guides.

chenopis commented:
@steveperry-53 Ok, I refactored it into two pages: container-lifecycle-hooks.md and container-environment.md. I also tracked down all the related links and updated them.

steveperry-53 commented on docs/concepts/workloads/containers/container-environment.md:
```diff
@@ -0,0 +1,63 @@
+{:toc}
+
+{% capture body %}
+## Container environment
```
Use title casing: Container Environment.

steveperry-53 commented on docs/concepts/workloads/containers/container-environment.md:
```diff
@@ -0,0 +1,63 @@
+{% capture body %}
+## Container environment
+
+The Kubernetes Container Environment provides several important resources to containers:
```
I don't think Container Environment is the proper name of anything in Kubernetes. So no caps.

steveperry-53 commented on docs/concepts/workloads/containers/container-environment.md:
```diff
@@ -0,0 +1,63 @@
+The Kubernetes Container Environment provides several important resources to containers:
+
+* A filesystem, which is a combination of an [image](/docs/concepts/workloads/containers/images) and one or more [volumes](/docs/concepts/storage/volumes).
+* Information about the container itself.
```
Throughout the doc, capitalize API objects like Container and Pod.

steveperry-53 commented on docs/concepts/workloads/containers/container-environment.md:
```diff
@@ -0,0 +1,63 @@
+Services have dedicated IP addresses and are available to the container via DNS,
+if [DNS addon](http://releases.k8s.io/{{page.githubbranch}}/cluster/addons/dns/) is enabled. 
+Because DNS is still not an enumerable protocol, however,
+environment variables are provided to enable Containers to do discovery.
```
Try to avoid phrases like "still non an enumerable protocol." If you have to talk about the present moment as opposed to the future, isolate those statements a dedicated section at the end of the doc. Also, make sure any of these present/future statements are verified by a tech reviewer.

steveperry-53 commented on docs/concepts/workloads/containers/container-lifecycle-hooks.md:
```diff
@@ -0,0 +1,121 @@
+
+`PostStart`
+
+This hook executes immediately after a container is created.
```
Caps for API objects like Containers and Pods.

steveperry-53 commented on docs/concepts/workloads/containers/container-lifecycle-hooks.md:
```diff
@@ -0,0 +1,121 @@
+### Hook handler implementations
+
+Containers can access a hook by implementing and registering a handler for that hook.
+There are two types of Hook handlers that can be implemented for Containers:
```
No caps for Hook.

steveperry-53 commented on docs/concepts/workloads/containers/container-lifecycle-hooks.md:
```diff
@@ -0,0 +1,121 @@
+
+Hook delivery is intended to be *at least once*,
+which means that a hook may be called multiple times for any given event,
+such as for `start` or `stop`.
```
start -> PostStart, stop -> PreStop

steveperry-53 commented on docs/concepts/workloads/containers/container-lifecycle-hooks.md:
```diff
@@ -0,0 +1,121 @@
+there is no attempt to resend.
+In some rare cases, however, double delivery may occur.
+For instance, if a kubelet restarts in the middle of sending a hook, 
+the hook may be resent after the kubelet comes back up.
```
may -> might

steveperry-53 commented on docs/concepts/workloads/containers/images.md:
```diff
@@ -0,0 +1,328 @@
+
+{% capture overview %}
+
+Each container in a pod has its own image.  Currently, the only type of image supported is a [Docker Image](https://docs.docker.com/engine/tutorials/dockerimages/).
```
I don't think Docker is the only supported image.

steveperry-53 commented:
@chenopis, Back to you. Comments in line.

chenopis commented:
@steveperry-53 Ok, thanks. I've addressed all of your comments.

chenopis commented:
@pwittrock Can you recommend an eng to do a tech review? Thx

pwittrock commented:
@Random-Liu Would be a good person.

chenopis commented:
Thanks @pwittrock

chenopis commented:
@Random-Liu Do you have time to do a tech review for this PR?

chenopis commented:
@Random-Liu Do you have time to review this? Thx
